src/zip/file.cr

```diff
@@ -0,0 +1,190 @@
+require "./file_info"
+
+# Provides random read access to zip file entries stores inside
+# a `File` or an `IO::Memory`.
+#
+# ### Example
+#
+# ```
+# Zip::File.open("./file.zip") do |file|
+#   # Iterate through all entries printing their filename and contents
+#   file.entries.each do |entry|
+#     p entry.filename
+#     entry.open do |io|
+#       p io.gets_to_end
+#     end
+#   end
+#
+#   # Random access to entries by filename is also provided
+#   entry = file["some_file.txt"]
+#   entry.open do |io|
+#     p io.gets_to_end
+#   end
+# end
+# ```
+class Zip::File
+  # Returns all entries inside this zip file.
+  getter entries : Array(Entry)
+
+  # Returns `true` if this zip file is closed.
+  getter? closed = false
+
+  # Returns the zip file comment.
+  getter comment = ""
+
+  # Opens a `Zip::File` for reading from the given *io*.
+  def initialize(@io : ::File | ::IO::Memory, @sync_close = false)
+    directory_end_offset = find_directory_end_offset
+    entries_size, directory_offset = read_directory_end(directory_end_offset)
+    @entries = Array(Entry).new(entries_size)
+    @entries_by_filename = {} of String => Entry
```

Comment:
I took inspiration from [Java's API](https://docs.oracle.com/javase/7/docs/api/java/util/zip/ZipFile.html#getEntry(java.lang.String)). Apparently [C#](https://msdn.microsoft.com/en-us/library/system.io.compression.ziparchive.getentry(v=vs.110).aspx) has it too. They don't mention the possibility of duplicated entries. Maybe it's so uncommon that it isn't worth considering, and a nicer API is better.